### PR TITLE
Improving the development experience

### PR DESCRIPTION
**Pull request by samuelcastro**
## What's the goal of this PR?

This PR includes a few libraries that will potentially improve the the development experience in general as well as create consistency in terms of code format and git commit messages

## [Prettier](https://prettier.io/):

Prettier is an opinionated code formatter that supports JavaScript (including ES2017), JSX, Flow, TypeScript, JSON, CSS, Less, SCSS, styled-components, Markdown, and GraphQL.
Prettier guarantees consistency in formatting by parsing your code into an abstract syntax tree (AST) and pretty-printing the AST.

### Motivations

When working with different teams, specially open sources projects, where contributors are usually dispersed, often strangers, and sometimes new to the project, using Prettier, individuals who are not familiar with the project’s style guidelines have one less thing to worry about when making a pull request. And owners, maintainers, and collaborators don’t have to enforce a style guide using nits in code reviews. At the same time, Prettier makes code formatting consistent across an entire codebase, increasing legibility.

**To use:**

`yarn prettier:fix`

## [Commitizen](http://commitizen.github.io/cz-cli)

Commitizen is a command line tool that allows developers create consistent commit messages. When you commit with commitizen, you'll be prompted to fill out any required commit fields at commit time. Contributors don't need to worried about what's the preferred format messages and so on.

### Motivations

Developers all have their own ways of formatting commit messages. One developer might put a short message and another might specify the type of changes made and include a github issue for example. These inconsistencies make it difficult to sift through the commit history. So using commitizen we will standardize our commit messages properly.

**To use:**

`yarn commit`

commitizen

## [Lint Staged](https://github.com/okonet/lint-staged)

Lint Staged is a library that will allows us run any given task (eslint or prettier) when new commit are created.

### Motivations

Linting makes more sense when run before committing your code. By doing so you can ensure no errors go into the repository and enforce code style. But running a lint process on a whole project is slow and linting results can be irrelevant. Ultimately you only want to lint files that will be committed. Lint Staged contains a script that will run arbitrary shell tasks with a list of staged files as an argument, filtered by a specified glob pattern.

**To use:**

Automatically on a a new git commit.

## Conclusion

With this setup will allow any new developer to contribute still making sure that we keep a good code standard and consistent and meaningful git messages, all automatically trigged with a simple git commit.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -120,6 +120,7 @@
   },
   "devDependencies": {
     "@next/bundle-analyzer": "9.4.1",
+    "@nordnet/cz-conventional-changelog": "^1.1.0",
```
Comment (Vadorequest) [inline package.json:124]: I advise you to `npm config set save-exact true` to use fixed versions by default. Mostly for a security/stability standpoint.

https://stackoverflow.com/questions/30656189/make-npm-install-save-add-a-strict-version-to-package-json

File: package.json [hunk 2]
```diff
@@ -36,12 +36,12 @@
     "doc:gem:install": "cd docs/ && bundle install",
     "lint": "esw src/ -w --ext .ts --ext .tsx",
     "lint:once": "eslint src/ --ext .ts --ext .tsx",
-    "lint:fix": "eslint src/ --ext .ts --ext .tsx --fix",
+    "lint:fix": "eslint src/ --ext .ts --ext .tsx --cache --fix",
```
Comment (Vadorequest) [inline package.json:39]: What does the --cache option do?

**Comment by Vadorequest** [comment]
Thanks for the PR, here are a quick feedback.

The reasons behind those changes are good, but it's a lot of changes to "swallow" (4k changes).

Don't get me wrong, it's a welcomed change, as I wanted to configure Prettier for the same reasons, but there are other related issues (I currently format code through WebStorm IDE, I'd need to change that, not necessarily complicated but I need to figure it out). Also, Prettier config seems to be light, and probably doesn't match eslint/WebStorm configs.

I'd have preferred we discuss this beforehand (I don't want anyone spending time on things that may not be merged), and probably do those changes in separate PRs ("lint stage" depends on the two, but could have been done afterwards).

Also, are those changes aimed at improving NRN itself, or projects built with NRN? In both case, additional documentation is needed, but not aimed at the same people. To me, Commitizen is mostly useful to people using NRN for their own projects, while Prettier and Lint staged are useful to both NRN maintainers and NRN-based projects.

**Comment by Vadorequest** [comment]
I've applied the prettier changes into one of my own branch to test the behaviour.

Overall, it's not working great at the moment.
There are rules that conflict between eslint and prettier.

For instance, I ran `yarn prettier:base --write src/components/pageLayouts/Footer.tsx` and I got this:

image

- `{t('footer.terms.link', "Conditions générales d'utilisation")}` got replaced but doesn't respect eslint rules

image

- Those changes are not really welcome, they don't improve readability IMHO.
- CSS changes are great, eslint doesn't handle them

Overall, my main issue is that I don't know how to configure prettier to match my code formatting. Someone else wrote about a similar concern (see https://github.com/prettier/prettier/issues/6069). The problem is that even though I can preview some changes using `yarn lint:fix:preview`, it only shows me part of the prettier errors/warnings. And it doesn't tell me **how I can update those rules**, unlike what eslint does.

image

All errors are tagged with `prettier/prettier` and I need to go through the whole doc to find rules that I should overwrite. It's such a waste of time. I don't know how many hours I need to spend on this to adapt prettier rules to the current code style.

**Comment by samuelcastro** [comment]
> Thanks for the PR, here are a quick feedback.
> 
> The reasons behind those changes are good, but it's a lot of changes to "swallow" (4k changes).
> 
> Don't get me wrong, it's a welcomed change, as I wanted to configure Prettier for the same reasons, but there are other related issues (I currently format code through WebStorm IDE, I'd need to change that, not necessarily complicated but I need to figure it out). Also, Prettier config seems to be light, and probably doesn't match eslint/WebStorm configs.

Using a code formatter that's a community driven solution rather than related to a specific brand (WebStorm) is really the best option, not everyone uses WebStorm.

> I'd have preferred we discuss this beforehand (I don't want anyone spending time on things that may not be merged), and probably do those changes in separate PRs ("lint stage" depends on the two, but could have been done afterwards).

It was not a waste of time for me considering that I'm using this setup anyway, so I felt it'd be good to share and yes next time I'll open up a an issue first to discuss the viability, same thing I did with the image and graphql optimization.

> Also, are those changes aimed at improving NRN itself, or projects built with NRN? In both case, additional documentation is needed, but not aimed at the same people. To me, Commitizen is mostly useful to people using NRN for their own projects, while Prettier and Lint staged are useful to both NRN maintainers and NRN-based projects.

From my perspective it'll help both kind of users, my motivations are described in the PR descriptions.

**Comment by samuelcastro** [comment]
> I've applied the prettier changes into one of my own branch to test the behaviour.
> 
> Overall, it's not working great at the moment.
> There are rules that conflict between eslint and prettier.
> 
> For instance, I ran yarn prettier:base --write src/components/pageLayouts/Footer.tsx and I got this:

With the conflict issues I've already included [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) which would ideally fix up those issues, not sure why you're still getting it, maybe I need to change the precedence on `.eslintrc`

Anyway, if you think it's not yet the time to bring those changes don't worry, let's close this and move on.

File: package.json (hunk 2 above)
Comment (samuelcastro) [reply package.json:39]: `--cache` only checks changed files, we don't need to check the whole project every time.

File: package.json (hunk 1 above)
Comment (samuelcastro) [reply package.json:124]: make sense.

**Comment by Vadorequest** [comment]
I agree, using Prettier (or similar) is better than using WebStorm, there is no doubt about that. My concern is around the usage migration from WebStorm to Prettier, because I'm use to it and need to learn how it'll work with Prettier.

Prettier is natively supported by WebStorm since very recently. https://prettier.io/docs/en/webstorm.html

I'm really looking towards using Prettier (or similar), but my main issue is that it doesn't explain why it does things, it just do them and I don't know what I should change to change the behaviour. Eslint is more developer-friendly about that.

**Comment by samuelcastro** [comment]
So what are we going to do? Close this for now?

**Comment by Vadorequest** [comment]
I won't/can't merge it as it for various reasons (https://github.com/UnlyEd/next-right-now/pull/52 would conflict a lot for example), and the way Prettier has been integrated doesn't fit well (lots of issues). I'm not sure what to do about it.

Committizen and lint staged are two good utilities. I'll see what how I can integrate those (lint staged makes less sense without Prettier)

**Pull request closed without merging**